lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -1060,6 +1086,80 @@ def _restart_opensearch(self, event: _RestartOpenSearch) -> None:
 
         self._start_opensearch_event.emit()
 
+    def _settle_voting_exclusions(self, unit_is_stopping: bool = False):  # noqa C901
+        """Settle the exclusions for all voting units."""
+        hosts = self.alt_hosts
+        if (
+            self.opensearch_peer_cm.deployment_desc().typ != DeploymentType.MAIN_ORCHESTRATOR
+            and (peer_cm_rel_data := self.opensearch_peer_cm.rel_data()) is not None
+        ):
+            # Also consider peer-relation units
+            hosts.extend([node.ip for node in peer_cm_rel_data.cm_nodes])
+
+        for attempt in Retrying(stop=stop_after_delay(300), wait=wait_fixed(10)):
+            with attempt:
+                nodes = ClusterTopology.nodes(self.opensearch, use_localhost=True, hosts=hosts)
+
+                node = None
+                for node in nodes:
+                    if node.name == self.unit_name:
+                        break
+                if unit_is_stopping and node:
+                    # This is a stop operation, we exclude the unit that is stopping
+                    # if still in the list
+                    nodes.remove(node)
+                    # we can finish the for loop here
+                    break
+                elif not unit_is_stopping and not node:
+                    # This unit is starting, we must assure it shows up in the list
+                    # Let's retry
+                    raise OpenSearchStartError("Node not in the list of nodes")
```

Comment:
Can you explain this? if we call `settle_voting_exclusions` from `post_start_init`, why should we have this check + exception?

Reply:
First, settle voting logic must be executed synchronously with a (re)start. This is why I am adding these two retry loops (one now, another at its end). Ideally, if we have a problem here, that may cause the entire cluster to get stuck in "metadata error" as we were seeing before.

Because between powering up and actually having the service up takes a while, this for loop assures we wait long enough for the opensearch service to come up.